### DISCUSSION TITLE
command.stamp hangs when using with uvcorn

### DISCUSSION THREAD

**Discussion by rajshrestha86** [sqlalchemy/alembic#1155] [Usage Questions]
**Describe the bug**
I am trying to stamp a revision from code in FastAPI. But when I am running `command.stamp` it hangs there and it doesn't start the API server. 

**Expected behavior**
The alembic should have stamp the db and have started the API server like below:
```
uvicorn app.main:app --port 8000
INFO:     Started server process [96686]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

**To Reproduce**
Here is a snippet of the code that I am working on:

```py
import logging

from alembic import command, config
from fastapi import FastAPI
from sqlmodel import create_engine

DATABASE_URI = "sqlite://"

logging.basicConfig(level=logging.INFO)
log = logging.getLogger(__name__)

app = FastAPI()

sqlmodel_engine = create_engine(DATABASE_URI)


with sqlmodel_engine.begin() as conn:
    alembic_cfg = config.Config("alembic.ini")
    alembic_cfg.attributes["connection"] = conn
    command.stamp(alembic_cfg, "head")
```

The app hangs with following output when running with uvicorn: `uvicorn app.main:app --port 8000`

**Output**
```
uvicorn app.main:app --port 8000
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
# Stuck after this
```

If I comment the line with `command.stamp` the API server starts successfully as in the first code snippet. If I run the code without uvicorn py just running `python main.py` it completes it's execution but when I start it with `uvcorn` it gets stuck. Can't seem to figure out what's the issue here.

**Comment by zzzeek** [comment]
what if you run a simple query instead of command.stamp, like `connection.execute(text("select 1"))`, does that return?

also:

1. what is the actual database URL and driver in use?   You seem to have a default of "sqlite://" memory DB, is that the actual backend you are running on ?
2. what SQLAlchemy version?

**Reply by rajshrestha86** [reply]
Yes it does. Running queries like `conn.execute("select 1")` starts the API server.

1. The DB_URI doesn't seem to affect  anything. I tried this on both `postgres` and `sqlite` with the same result. 
2. I am using `sqlmodel` and here are the version details:
- uvicorn==0.20.0
- fastapi==0.88.0
- sqlalchemy==1.4.41 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4, 3.5'
- sqlmodel==0.0.8

**Reply by zzzeek** [reply]
it would seem then that your env.py is either 1. not using the "connection" that you are placing in "attributes" and it's hung on trying to connect with a different engine or 2. is getting the connection but then calling some command causing the hang

**Reply by rajshrestha86** [reply]
I also looked into `env.py` to make sure that I am re-using the connection sent in attributes as mentioned here: https://alembic.sqlalchemy.org/en/latest/cookbook.html#connection-sharing If I run the file with just `python main.py` it doesn't hang.

**Reply by zzzeek** [reply]
can you go into env.py and try to reduce the steps within to isolate where the hang occurs?  such as change:

```
        with context.begin_transaction():
            context.run_migrations()
```

to 

```
        with context.begin_transaction():
           pass
```

things like that.   When running against Postgresql, do you see more than one connection in play?   a "hang" can be a lot of things but with sqlite memory DB there's really nothing that can "hang".     Can you run a command like "command.history()" instead, is there something limiting filesystem access?

**Reply by zzzeek** [reply]
finally, this is apparently in stdlib now so this makes it much easier, run [faulthandler](https://docs.python.org/3/library/faulthandler.html).  Add this to your program

```py
import faulthandler
import signal
faulthandler.register(signal.SIGUSR1)
```

run the program, get the hang.  get the pid of the Python process that's hanging.  then do a "kill -SIGUSR1 <pid>".  you should get a stack trace from the Python program.

**Reply by rajshrestha86** [reply]
It hangs even though I changed `run_migrations` to `pass`:
```
with context.begin_transaction():
    context.run_migrations()
```
to 
```
 with context.begin_transaction():
           pass
```

Also, added faulthandler and ran: `python main.py faulthandler` and it still hangs without any errors.
```
import faulthandler
import signal
faulthandler.register(signal.SIGUSR1)

if __name__ == '__main__':
    uvicorn.run("main:app", port=5000, log_level="info")
```

**Reply by rajshrestha86** [reply]
What's interesting is that if I use Flask instead of FastAPI then it works,  doesn't hang. But not sure why it hangs when using with `FastAPI` and `uvicorn`.

**Reply by zzzeek** [reply]
OK....using "faulthandler" alone is not enough, as mentioned, when it hangs, **keep it hung, then go to your console** and do a signal using ``kill -SIGUSER1 <pid>`` - that's what triggers faulthandler to print  stack where it's hung

**Reply by bavarian-ng** [reply]
In our case, we had the same problem... I removed the logging options from the alembic.ini as well as the call for loading it from env.py as it seemed to interfere with each other if there were different log levels set in our uvicorn application and alembic.
Maybe this helps someone with same error in the future. :)

**Reply by dunielpls** [reply]
I came across this issue too, thank you @bavarian-ng. Have anyone managed to figure out why `logging.config.fileConfig(config.config_file_name)` causes this behaviour?

**Reply by dunielpls** [reply]
This is caused by `fileConfig`'s argument `disable_existing_loggers` defaulting to a value of `True`. Changing the relevant line in `env.py` to `fileConfig(config.config_file_name, disable_existing_loggers=False)` solved my issue.
